src/ressa/index.rs

```diff
@@ -0,0 +1,166 @@
+        self.ast_languages.insert(IndexableKey::new(node), langs);
+    }
+}
+
+/// Descriptive alias. Also, centralizes size; may need to expand in the future.
+pub type LanguageSet = Bitmap<32>;
+
+/// Retrieve whether a language is recorded
+pub fn langset_contains(langset: &LanguageSet, lang: &Language) -> bool {
+    langset.get(Language::get_index(lang))
+}
+
+/// Record a language to the set
+pub fn langset_set(langset: &mut LanguageSet, lang: &Language) -> bool {
+    langset.set(Language::get_index(lang), true)
+}
+
+/// Compute the languages to index over
+pub fn compute_index_languages<'a>(
+    patterns: &Vec<NodePattern>,
+    nodes: Vec<IndexableEntry<'a>>,
+) -> LaastIndex<'a> {
+    // Compute all languages to index on
+    let mut indices = HashMap::new();
+    for pattern in patterns.iter() {
+        let language = pattern.get_language();
+        if !indices.contains_key(&language) {
+            indices.insert(language, vec![]);
+        }
+    }
+
+    // Generate indices
+    let mut indices = LaastIndex::new(indices, HashMap::new());
+    for node in nodes {
+        // println!("Node: {:#?}", node.get_language());
+        index(Language::Unknown, node, LanguageSet::new(), &mut indices);
+    }
+
+    indices
+}
+
+/// Run the indexing procedure over the given node
+fn index<'a>(
+    mut current_lang: Language,
+    current: IndexableEntry<'a>,
+    mut curr_langs: LanguageSet,
+    indices: &mut LaastIndex<'a>,
+) -> LanguageSet {
+    // If language changed, update data
+    let new_lang = current.get_language();
+    if new_lang != current_lang {
+        // If it's a new language in this subtree, try to create an index entry
+        if !langset_contains(&curr_langs, &new_lang) {
+            // println!("Index {:#?}", new_lang);
+            indices.add_if_valid(current);
+        }
+
+        // Update current language data
+        langset_set(&mut curr_langs, &new_lang);
+        current_lang = new_lang;
+    }
+
+    // Visit decendents and retrieve descendent languages
+    let mut my_set = LanguageSet::new();
+    langset_set(&mut my_set, &current_lang);
+    for node in current.get_children() {
+        my_set.bitor_assign(index(current_lang, node, curr_langs.clone(), indices));
+    }
+
+    // Record and return node's languages
+    indices.record_subtree_langs(current, my_set);
+    my_set
+}
+
+pub trait Indexable: RessaNodeExplorer + NodeLanguage + ChildFields + std::fmt::Debug {
+    fn get_children(&self) -> Vec<&dyn Indexable>;
+}
+
+pub trait ChildFields {
+    fn get_fields(&self) -> Vec<Vec<&dyn Indexable>>;
+}
+
+impl<T> Indexable for T
+where
+    T: RessaNodeExplorer + ChildFields + NodeLanguage + std::fmt::Debug,
+{
+    fn get_children(&self) -> Vec<&dyn Indexable> {
+        self.get_fields()
+            .into_iter()
+            .flat_map(|field_children| field_children)
+            .collect()
+    }
+}
```

Comment:
Oh the blanket impl is still here. What was the other blanket impl you were referring to?

Reply:
Blanket implementation of `explore`, see my earlier comment.